classes/ThankYous/ThankYousRepository.php

```diff
@@ -0,0 +1,1121 @@
+			}
+
+			try
+			{
+				//TODO: Replace with a non-static post People API update
+				$user_image_url = User::GetPhotoUrl($user->GetId());
+			} catch (CDNSystemException $cdn_system_exception)
+			{
+				$this->logger->error("Failed to Get User's Photo URL when Creating Thankable: " . $cdn_system_exception->getMessage());
+				$user_image_url = null;
+			}
+
+			//TODO: Replace with a non-static post People API update
+			$user_profile_url = User::GetProfileUrl($user->GetId(), false);
+
+			$users[$user_offset] = $this->thankable_factory->Create($user->GetFullname(), $user->GetId(), $owner_class_id, $user->GetExAreaId(), $user_image_url, $user_profile_url);
+		}
+
+		return $users;
+	}
+
+	/**
+	 * Saves a Thank You to the repository. If the Thank You is new the ID its ID will also be set.
+	 *
+	 * @param ThankYou $thank_you
+	 * @return int ID of saved Thank You
+	 * @throws ThankYouNotFound - If the Thank You could not be found in the Repository.
+	 */
+	public function Save(ThankYou $thank_you)
+	{
+		$id = $thank_you->GetId();
+		if (isset($id))
+		{
+			$thank_yous = $this->GetThankYous([$id]);
+			if (!isset($thank_yous[$id]))
+			{
+				throw new ThankYouNotFound("Failed to Save Update to Thank You, Thank You not found");
+			}
+		}
+
+		$author_id = $thank_you->GetAuthor()->GetId();
+
+		$date_created = clone $thank_you->GetDateCreated();
+		$date_created->setTimezone(new DateTimeZone("UTC"));
+		$date_created_string = $date_created->format('YmdHis');
+
+		$description = $thank_you->GetDescription();
+
+		$id = $this->SaveThankYou($author_id, $date_created_string, $description, $id);
+
+		$thank_you->SetId($id);
+
+		$thankees = $thank_you->GetThankable();
+		if (isset($thankees))
+		{
+			$this->DeleteThankYouThankees($id);
+
+			foreach ($thankees as $thankable)
+			{
+				$owner_class_id      = $thankable->GetOwnerClass();
+				$owner_class_item_id = $thankable->GetId();
+
+				if (isset($owner_class_id) && isset($owner_class_item_id))
+				{
+					$this->SaveThankYouThankee($id, $owner_class_id, $owner_class_item_id);
+				}
+			}
+		}
+
+		$users = $thank_you->GetUsers();
+		if (isset($users))
+		{
+			$this->DeleteThankYouUsers($id);
+
+			foreach ($users as $user)
+			{
+				$user_id = $user->GetId();
+
+				if (isset($user_id))
+				{
+					$this->SaveThankYouUser($id, $user_id);
+				}
+			}
+		}
+
+		return $id;
+	}
+
+	/**
+	 * @param int $id
+	 */
+	public function Delete(int $id)
+	{
+		$this->DeleteThankYouUsers($id);
+		$this->DeleteThankYouThankees($id);
+		$this->DeleteThankYou($id);
+	}
+
+	/**
+	 * Saves a Thank You. If an ID is provided, an existing record will be updated, if not a new entry will be
+	 * created.
+	 *
+	 * @param int      $author_id
+	 * @param int      $date_created
+	 * @param string   $description
+	 * @param int|null $id
+	 * @return int
+	 */
+	private function SaveThankYou(int $author_id, int $date_created, string $description, ?int $id)
```

Comment:
Why doesn't this just accept a thank you object? The complexity of formatting dates for the database could then be moved to this method.

Same for `SaveThankYouThankee()`, in terms of accepting an object.

Reply:
Fixed in 8433cddc28d54b3866cfe5ca20a75918bb705c8c and e194680b08ea2aa7560b059f6ec026a4af45844e

Definitely worth a review!

Reply:
Not sure why the new throws are necessary here, these kind of checks should be done at the API level instead of the repo level. In the repo you could simply cast the values to the types they need to be.

Reply:
Fixed in e8ba803877d9b20349ea3d0e7a91274172381dd8